### PR TITLE
feat(api): add recaptcha verification on user creation

### PR DESCRIPTION
**Pull request by holahula**
## Summary
Add a new service to support recaptcha verification on user creation following the instructions [here](https://developers.google.com/recaptcha/docs/verify#api-request) . This PR builds on the work in #1206. 

TODO: `RECAPTCHA_SECRET_KEY` is must be added to the production config

## Testing Plan
- Added unit tests for the new recaptcha verification service
- Refactored existing user service tests to mock recaptcha verification for user creation and added tests to cover failed recaptcha verification

## Breaking Change

Is this a breaking change? If yes, add notes below on why this is breaking and
what additional work is required, if any.

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: src/recaptcha-verification/recaptcha-verification.service.spec.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+  describe('verify', () => {
+    it('should return true if the recaptcha verification is successful', async () => {
+      const verifyRecaptcha = jest.spyOn(
+        recaptchaVerificationService,
+        'verify',
+      );
+      const response = await recaptchaVerificationService.verify('token');
+      expect(verifyRecaptcha).toHaveBeenCalled();
+      assert.ok(verifyRecaptcha.mock.calls);
```
Comment (rolltable) [inline src/recaptcha-verification/recaptcha-verification.service.spec.ts:33]: this is not needed because of line 32 is testing the same thing

Comment (NullSoldier) [inline src/recaptcha-verification/recaptcha-verification.service.spec.ts:33]: this is duplicating line 32

File: src/recaptcha-verification/recaptcha-verification.service.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
+import { ApiConfigService } from '../api-config/api-config.service';
+
+@Injectable()
+export class RecaptchaVerificationService {
+  constructor(private readonly configService: ApiConfigService) {}
+
+  async verify(recaptcha?: string | undefined): Promise<boolean> {
```
Comment (NullSoldier) [inline src/recaptcha-verification/recaptcha-verification.service.ts:12]: you should just make this required and not undefined

File: src/recaptcha-verification/recaptcha-verification.service.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    const recaptchaSecret = this.configService.get<string>(
+      'RECAPTCHA_SECRET_KEY',
+    );
+
+    const recaptchaVerificationUrl = `https://www.google.com/recaptcha/api/siteverify?secret=${recaptchaSecret}&response=${recaptcha}`;
+
+    const isVerified = await fetch(recaptchaVerificationUrl)
+      .then((response) => response.json())
```
Comment (NullSoldier) [inline src/recaptcha-verification/recaptcha-verification.service.ts:27]: If you're doing an await, you don't need to put any of this code into the then(). you can just move it out into the main code area.

const response = await fetch() 
const body = response.json() as {{
          success: boolean;
          challenge_ts: Date;
          hostname: string;
          'error-codes'?: string[];
        }

File: src/recaptcha-verification/recaptcha-verification.service.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        "Missing 'recaptcha' field",
+        HttpStatus.BAD_REQUEST,
+      );
+    }
+
+    const recaptchaSecret = this.configService.get<string>(
+      'RECAPTCHA_SECRET_KEY',
+    );
+
+    const recaptchaVerificationUrl = `https://www.google.com/recaptcha/api/siteverify?secret=${recaptchaSecret}&response=${recaptcha}`;
```
Comment (NullSoldier) [inline src/recaptcha-verification/recaptcha-verification.service.ts:24]: Can we provide the users IP address here?

File: src/recaptcha-verification/recaptcha-verification.service.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        (response: {
+          success: boolean;
+          challenge_ts: Date;
+          hostname: string;
+          'error-codes'?: string[];
+        }) => {
+          if (response.success) {
+            return true;
+          }
+          return false;
```
Comment (NullSoldier) [inline src/recaptcha-verification/recaptcha-verification.service.ts:38]: this is already a boolean, you don't need an if statement.

```suggestion
          return response.success;
```

File: src/users/users.controller.spec.ts [hunk 6]
```diff
@@ -566,6 +574,8 @@ describe('UsersController', () => {
           })
           .expect(HttpStatus.CREATED);
 
+        expect(recaptchaVerification).toHaveBeenCalledTimes(1);
+        assert.ok(recaptchaVerification.mock.calls);
```
Comment (NullSoldier) [inline src/users/users.controller.spec.ts:578]: This duplicates the above expect

File: src/users/users.controller.ts [hunk 7]
```diff
@@ -298,6 +301,17 @@ export class UsersController {
...
     dto: CreateUserDto,
   ): Promise<User> {
+    const isRecaptchaValid = await this.recaptchaVerificationService.verify(
+      dto.recaptcha,
+    );
+
+    if (!isRecaptchaValid) {
+      throw new HttpException(
+        'ReCAPTCHA verification failed',
+        HttpStatus.UNAUTHORIZED,
```
Comment (NullSoldier) [inline src/users/users.controller.ts:311]: Don't use unauthorized for this, return validation failed. Unauthorized is for when things like passport auth fail.

File: src/recaptcha-verification/recaptcha-verification.service.ts (hunk 5 above)
Comment (holahula) [reply src/recaptcha-verification/recaptcha-verification.service.ts:38]: fixed

**Comment by holahula** [comment]
@rohanjadvani There was a [test failure ](https://github.com/iron-fish/ironfish-api/actions/runs/4127107634/jobs/7129900845#step:10:89) on the first run but succeeded after re-running the jobs. Is this related to the changes in the PR?

File: src/recaptcha-verification/recaptcha-verification.service.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+      'RECAPTCHA_SECRET_KEY',
+    );
+
+    const recaptchaBaseUrl = 'https://www.google.com/recaptcha/api/siteverify';
+    const recaptchaVerificationUrl = `${recaptchaBaseUrl}?secret=${recaptchaSecret}&response=${recaptcha}`;
+
+    try {
+      const isVerified = await axios
+        .post<GoogleRecaptchaVerificationJsonResponse>(recaptchaVerificationUrl)
+        .then(({ data }) => {
```
Comment (rohanjadvani) [inline src/recaptcha-verification/recaptcha-verification.service.ts:23]: Is this `.then` needed? Or could we just return `response.data.success`?

File: src/users/users.controller.spec.ts [hunk 9]
```diff
@@ -574,6 +583,58 @@ describe('UsersController', () => {
           discord,
         });
       });
+
+      describe('recaptcha verification failure', () => {
+        afterEach(() => {
+          jest.clearAllMocks();
+        });
+
+        it('return 400 on missing recaptcha token', async () => {
```
Comment (rohanjadvani) [inline src/users/users.controller.spec.ts:592]:
```suggestion
        it('returns 400 on missing recaptcha token', async () => {
```

File: src/users/users.controller.ts [hunk 10]
```diff
@@ -298,6 +301,23 @@ export class UsersController {
     )
     dto: CreateUserDto,
   ): Promise<User> {
+    if (dto.recaptcha === undefined) {
```
Comment (rohanjadvani) [inline src/users/users.controller.ts:304]: Can we just update the DTO to check that this is defined?

File: src/recaptcha-verification/recaptcha-verification.service.ts (hunk 8 above)
Comment (holahula) [reply src/recaptcha-verification/recaptcha-verification.service.ts:23]: `response.data.success` works - updated

File: src/users/interfaces/create-user-options.ts [hunk 11]
```diff
@@ -0,0 +1,11 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+export interface CreateUserOptions {
+  country_code: string;
```
Comment (rohanjadvani) [inline src/users/interfaces/create-user-options.ts:5]: We use camelCase for internal options / types
```suggestion
  countryCode: string;
```

File: src/users/users.controller.ts [hunk 12]
```diff
@@ -298,7 +301,25 @@ export class UsersController {
...
+      throw new HttpException(
+        'ReCAPTCHA verification failed',
+        HttpStatus.UNPROCESSABLE_ENTITY,
+      );
+    }
+
+    return this.usersService.create({
+      email: dto.email,
+      graffiti: dto.graffiti,
+      country_code: dto.country_code,
```
Comment (rohanjadvani) [inline src/users/users.controller.ts:318]:
```suggestion
      countryCode: dto.country_code,
```

**Review by rohanjadvani — APPROVED** [review]

**Review by NullSoldier — APPROVED** [review]

File: src/telemetry/telemetry.controller.ts [hunk 13]
```diff
@@ -147,9 +148,8 @@ export class TelemetryController {
     }
 
     await this.nodeUptimes.addUptime(user);
-    if (ipAddress) {
-      await this.usersService.updateHashedIpAddress(user, ipAddress);
-    }
+
+    await this.usersService.updateHashedIpAddress(user, ipAddress);
```
Comment (holahula) [inline src/telemetry/telemetry.controller.ts:152]:
```suggestion
if (ipAddress !== request.ip) {
    await this.usersService.updateHashedIpAddress(user, ipAddress);
    }
```

**Pull request merged**